python/simulation.py

```diff
@@ -859,6 +864,155 @@ def amplitude(self, point, component):
         return mp.eigenmode_amplitude(self.swigobj, swig_point, component)
 
 
+class Pade:
+    """
+    Pade is implemented as a class with a [`__call__`](#Pade.__call__) method,
+    which allows it to be used as a step function that collects field data from a given
+    point and runs [Pade](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.pade.html)
+    on that data to extract an analytic rational function which models the frequency response.
+
+    See [`__init__`](#Pade.__init__) for details about constructing a `Pade`.
+
+    In particular, Pade stores the discrete time series $\\hat{f}[n]$ corresponding to the given field
+    component as a function of time and expresses it as:
+
+    $$\\hat{f}(\\omega) = \\sum_n \\hat{f}[n] e^{i\\omega n \\Delta t}$$
+
+    The above is a "Taylor-like" polynomial in $$n$$ with a Fourier basis and
+    coefficients which are the sampled field data. We then compute the Pade approximant
+    to the analytic form of this function as:
+
+    $$R(\\omega) = \\frac{P(\\omega)}{Q(\\omega)}$$
+
+    Where $$P$$ and $$Q$$ are polynomials of degree $$m$$ and $$n$$, and $$m + n + 1$$ is the
+    degree of agreement of the Pade Approximant to the analytic function $$f(\\omega)$$. This
+    function $$R$$ is stored in the callable method `pade_instance.freq_response`.
+    Be sure to save a reference to the `Pade` instance if you wish
+    to use the results after the simulation:
+
+    ```py
+    sim = mp.Simulation(...)
+    p = mp.Pade(...)
+    sim.run(p, until=time)
+    # do something with p.freq_response
+    ```
+    """
+
+    def __init__(
+        self,
+        c: int = None,
+        pt: Vector3Type = None,
```

Comment:
Any reason we can't do a `Volume` here instead?

Reply:
Yep there is certainly a tradeoff with storage. But it's useful to be able to do this. One could simply pass a point volume if that's all they care about. 

For adjoint sensitivities, we'll have to do this across the entire design region.

Reply:
This seems like a good feature to have. 

Your return value would then be a lambda that returns a numpy array of the spatial data (exactly in the shape of the `get_array` data.)

Reply:
I have a quick and dirty
```python
p_arr = []
q_arr = []
for k in range(np.shape(samples)[-1]):
     p,q = pade(samples[:,k], self.m, self.n)
     p_arr.append(p)
     q_arr.append(q)
```
which works fine if the sample data is 2-dimensional (time x 1 spatial dimension), but not if the sample is higher dimensional. I tried using `np.apply_along_axis()` but got ragged array issues when the length of the numerator `poly1d` was different from the length of the denominator `poly1d`. Any insight on how to do this efficiently (hopefully not just a triple for loop over the volume)?